fix a bug by adding a missing component import

foladipo opened the pull request:
Resolves #2802.

This PR fixes a bug that stopped the search modal from functioning properly. In summary, the bug was due to an improper import/export of a certain component.

### Test Instructions
- Make sure that your instance of RC has multiple users and orders.
- Log in as admin.
- Open the search modal using the search icon in the navigation bar.
- Enter a simple search query and check its results in either the `ACCOUNTS` or `ORDERS` tabs of the search results section.
- Open your browser console. Observe that the errors reported in #2802 are no longer there.

zenweasel commented:
@foladipo I no longer get an error but clicking on any of the result doesn't do anything. You might want to check with @joykare. I'm not sure that reverting to the `SortableTableLegacy` was the correct approach.

foladipo commented:
Ok @zenweasel . That has been fixed and the accounts/orders search results are now clickable.

zenweasel commented:
@rymorgan Random question but now that we have Order Search in the dashboard, should we eliminate it here?

zenweasel commented:
When I search for Accounts and click on a results this is what I see:

![dashboard_accounts](https://user-images.githubusercontent.com/72819/30466180-84454d16-9a0e-11e7-8984-8b012922459c.png)

rymorgan commented:
@zenweasel Yes, I think we can eliminate ordres search from the customer facing search. That's where I want to go with the accounts/user panel as well.

zenweasel commented:
@foladipo What's the status on this?

foladipo commented:
@zenweasel I fixed the latest issue you raised above, but it brought up another problem. So I'm busy trying to fix that right now. I should finish it before EOD.

foladipo commented:
@zenweasel I've fixed the mixed up actionView that you see when you click an accounts search result.

(cc @rymorgan) I have also removed the orders search tab from the general, customer-facing search.

zenweasel commented:
This is still not working. The account search results seem to get "stuck" at some point.

### Steps to replicate

1. Login as admin
1. Go to account search and search for your admin name
1. Click on the result and observe the accounts screen opens up for that user
1. Click on the Search Icon again and select "Accounts"
1. Observe that no matter what you put in there the results never change.

zenweasel commented:
So what did we decide as far as what should happen when you click on an account here? This is just opening the account panel for the logged in account, not for the user I searched for and clicked on which makes this pretty useless.

This might be beyond the scope of what @foladipo is trying to fix here though.

zenweasel commented:
This comment from @rymorgan from another PR that you were working on but closed

> If this ticket is for non-admin users, then we can send them to the user page. This will make sense in the future when we update the user profile page and also have a custom view when an admin is looking at a customer (aka a user with no permissions).
> 
> If the user is admin we need to add a card to edit that specific users' group and reset a password.

This issue (https://github.com/reactioncommerce/reaction/issues/2625) is open for that and I wonder if we should just fix that here. Right now the Account Search action when you click on the results are sort of useless.

I also opened https://github.com/reactioncommerce/reaction/issues/2903 which I think is the long term solution for this and seems like we should do sooner rather than later since I think this is the 3rd or 4th time we have revisited the account search which was never meant to be a long-term plan.

foladipo commented:
@zenweasel I agree that #2903 is probably the long term solution, and we will have to do it eventually. But due to the amount of consultation, design etc that that path will require, I agree with linking accounts search results to that user's profile page for now. Doing this will also resolve #2625.

As for the search getting stuck, I will look into it too.

aaronjudd commented:
@foladipo @zenweasel what's the status on this? this is a marketplace blocker...

foladipo commented:
@aaronjudd I'm working on a fix now. Should have an update by EOD.

rymorgan commented:
@foladipo per your question, an admin who searches for a user, should see all the cards their permissions group has access to. An owner for instance should everything.

foladipo commented:
@aaronjudd Sorry, I underestimated the scope of what needs to be done. I just pushed an update now, but it's the first part of what @rymorgan , @zenweasel and I agreed would complete this ticket. I now think getting this done might take about two more days.

foladipo commented:
*UPDATE*

Authorized users (e.g the admin) can now update the password of another user apart from themselves.

aaronjudd commented:
We should consider moving this PR back to [WIP] or close it, as there has been no real progress and currently in conflict. @zenweasel to review.

foladipo commented:
@impactmass We need your input on what parts of a user's profile an admin should be able to edit: is it their email, password or address?

foladipo commented:
Resolved conflict by removing code in `folusho-fix-orders-and-accounts-search-2802` that @spencern removed in [this commit](https://github.com/reactioncommerce/reaction/commit/f23079cd9789ecdf83b71d35e74f8b799f1ade86#diff-473965201c634c2840dc5cac9bfb3486).

impactmass commented:
@foladipo Re: your question. At this point, I'll actually say we leave out editing another user's data out of this ticket. View only is fine for now. The specific items that can be edited will (to me) require a bit more thought/planning. For example, there is/are argument(s) against having the right to edit a user's email, knowing that it is linked to a user being able to log in.

In short, I'm saying, on clicking a user link from search, route to the profile page with details that can benefit the admin (but not edit rights, yet). And also, that frees up this ticket from taking too much at once. Further discussion can go on in #2903

foladipo commented:
Thanks @impactmass . Consequently, I propose that when an admin searches for a user, s/he should be able to see just the user's email, name, id and last 10 orders. In that case, we might need a design from @rymorgan . Or should I come up with something simple?

rymorgan commented:
@foladipo We should show the admin the user's profile page with, username, email and order history. None of that should be editable as we should deal with this in the future with a little more planning, per @impactmass comment.

foladipo commented:
@rymorgan Roger that.

rymorgan commented:
You can use the same order history we have now on the user profile page: 
![screen shot 2017-10-03 at 11 24 37 am](https://user-images.githubusercontent.com/20409254/31141904-9989924a-a82d-11e7-89c7-7b72a2a50de3.png)

zenweasel commented:
@foladipo What's the status on this? Can we try to get this wrapped by EOD today?

foladipo commented:
@zenweasel I think we can, and I'm really working towards that.

foladipo commented:
@zenweasel @rymorgan I've finished implementing all we discussed and agreed on. I've also updated the test instructions above. So, please review. Will be waiting for your feedback.

foladipo commented:
### *For future readers*:
I have removed the code that allowed privileged users to modify another user's email. That code was added in `2e7e2ba`, but those changes have now been reverted in `341c3cb`, except for those in `server/api/core/accounts/password.js`. So if you want to do re-add this feature in the future (pun is coincidental), then consulting `2e7e2ba` might be a good starting point.

machikoyasuda commented:
Thanks for adding comments in https://github.com/reactioncommerce/reaction/pull/2834/files#diff-1d795a38de017c04a87f279e0a655c02. Looks great. Please also do that for https://github.com/reactioncommerce/reaction/pull/2834/files#diff-b1694abc4c64684f0d34d011b923ab10 file `imports/plugins/core/accounts/client/templates/profile/profile.js`

How to guide here: https://github.com/reactioncommerce/reaction-jsdoc/blob/master/README.md#how-to-write-docs

rymorgan commented:
A few things. When I create a user in a group like a store manager, their name does not show up in search. (it does on the user profile). 
Some things are missed labels. 
<img width="1411" alt="account_profile" src="https://user-images.githubusercontent.com/20409254/31189501-4a4b3010-a8ed-11e7-89f6-5393921fb78c.png">

rymorgan commented:
Also, when I place an order as registered user, even when it has been processed and completed by an admin, I don't see it showing up on the user profile page. 

<img width="801" alt="account_profile-1" src="https://user-images.githubusercontent.com/20409254/31189718-e155f4c2-a8ed-11e7-8b0f-95cecb00e6b2.png">

foladipo commented:
Noted @rymorgan . I'll fix the UI issues and then try to reproduce the orders bug. Be back in a jiffy.

foladipo commented:
@rymorgan The issues you raised have been fixed. Please check and let me know what else you find.

foladipo commented:
By the way, @zenweasel , I've had to modify two publications in fixing this bug. They are [this](https://github.com/reactioncommerce/reaction/commit/597e6888dceb8f569caca9b63d6aaac5cf53f5ae#diff-ed8a93fb7c829abae7924f5db470215d) and [this](https://github.com/reactioncommerce/reaction/pull/2834/commits/d714c7d3f7a9e20ade4c67d553da6d46d9fff4ef). I wonder if there are ramifications to these actions that I don't know of?

rymorgan commented:
Looks like my issues are fixed. Can you change the placeholder to be 'Not available' -- no parenthesis needed.

zenweasel commented on imports/plugins/core/accounts/client/templates/profile/profile.js:
```diff
@@ -7,6 +7,19 @@ import * as Collections from "/lib/collections";
 import { Components } from "@reactioncommerce/reaction-components";
 
 
+function isOwnerOfProfile() {
```
Needs JSDoc

zenweasel commented on imports/plugins/included/ui-search/lib/helpers/accountsTable.js:
```diff
@@ -28,7 +28,7 @@ export default function accountsTable() {
           return rowData.profile.firstName;
         }
-        return undefined;
+        return "(unvailable)";
```
Do we need i18n here?

zenweasel commented on server/api/core/accounts/password.js:
```diff
@@ -34,7 +34,7 @@ export function sendResetPasswordEmail(userId, optionalEmail) {
 
   // make sure we have a valid email
-  if (!email || !Array.includes(_.map(user.emails || [], "address"), email)) {
+  if (!email || !((_.map(user.emails || [], "address")).includes(email))) {
```
We shouldn't be using lodash `_.map` there. Same with the usages below.

machikoyasuda commented on private/data/i18n/en.json:
```diff
@@ -297,6 +297,10 @@
         "weibo": "Weibo"
       },
       "accountsUI": {
+        "yourOrders": "Your Orders",
```
Use sentence casing here: `Your orders`, `User's orders`

machikoyasuda commented:
JSdoc looks good ✔️

zenweasel commented:
I don't know exactly how this should work re: marketplace but when I invite a user to be a merchant shop owner I don't see their account show up in the search. thoughts @spencern?

zenweasel commented:
So it looks like results are still getting "stuck" here (notice there is no search term entered)
This didn't happen right away but what I have done so far:
1. Search for "brent" (and see the one result of my default owner login)
1. Invite a shop owner from the Marketplace settings
1. Invite a shop owner from the Account panel
1. Go back to account panel
1. That search result will not go away. If I search for another account that account will show up as well but the first one never leaves.

![account_profile](https://user-images.githubusercontent.com/72819/31264489-14ccfa30-aa9c-11e7-810f-d02fa28cd3e3.png)

![account_profile](https://user-images.githubusercontent.com/72819/31264558-8ea91c80-aa9c-11e7-8041-2452427b9841.png)

zenweasel commented:
So if I login as an invited shop owner I can see results for the owner and it also included order information
![account_profile](https://user-images.githubusercontent.com/72819/31264696-5e61a19a-aa9d-11e7-80fc-01764cbdae53.png)

zenweasel commented:
It just seems like results keep getting stuck. I searched for "bark" in the last search and then searched for "da" and got both the new results and the results from the last.

![account_profile](https://user-images.githubusercontent.com/72819/31265533-ddacf608-aaa1-11e7-9c0c-c3696e15c64c.png)

joykare commented:
I think you meant to add him as a reviewer rather than assign him :)

zenweasel commented:
I had told @foladipo that we should assign reviewers because that was what we were doing for a minute, but no longer

spencern commented:
@foladipo What's the status on this PR? I'd like to get it merged today, but currently has failing tests   and un-resolved review comments.

foladipo commented:
@spencern This PR's status is that it's ready for final review/merging. There were two outstanding review comments, and they've been fixed i.e:

1. that the search results were getting stuck, and that has been resolved by @joykare .

1. some concern about whether other admins should be able to see the owner's order history. @impactmass , @zenweasel , you and I have discussed that and concluded that it isn't.

About the failing test in CircleCI's build, I think it's a general problem in `marketplace` that was introduced in some other PR. I also see that you've tried to solve it [here](https://github.com/reactioncommerce/reaction/commit/8f4e447dc2d35478a1165401a2633e2b15b77552).

So I think we're ready for final review/merging.

spencern commented on imports/plugins/core/ui/client/containers/avatar.js:
```diff
@@ -1,11 +1,21 @@
-  const user = Accounts.findOne(Meteor.userId());
-  const email = user.emails[0].address;
+  const targetUserId = Reaction.Router.getQueryParam("userId");
+  let user = Accounts.findOne(targetUserId);
```
To make it easier for others to understand code that relates to accounts vs users, I'd like to see us use `account` whenever we're getting an account instead of `user`

foladipo replied:
Ok. I agree that that would probably be a better practice.

spencern commented on imports/plugins/included/ui-search/lib/containers/searchSubscription.js:
```diff
@@ -0,0 +1,73 @@
+  for (const product of productResults) {
+    if (product.hashtags) {
+      for (const hashtag of product.hashtags) {
+        if (!_.includes(hashtags, hashtag)) {
```
we should use the native `Array.prototype.includes` instead of lodash here.

foladipo replied:
That's true. This old use of lodash should have been updated.

spencern commented on imports/plugins/included/ui-search/lib/containers/searchSubscription.js:
```diff
@@ -0,0 +1,73 @@
+
+function getProductHashtags(productResults) {
+  const hashtags = [];
+  for (const product of productResults) {
```
this nested loop would be much better as an `Array.prototype.reduce`

foladipo replied:
True, using a `reduce` would have been better.

spencern commented on server/api/core/accounts/password.js:
```diff
@@ -34,7 +34,7 @@ export function sendResetPasswordEmail(userId, optionalEmail) {
 
   // make sure we have a valid email
-  if (!email || !Array.includes(_.map(user.emails || [], "address"), email)) {
+  if (!email || !user.emails || !user.emails.map(mailInfo => mailInfo.address).includes(email)) {
```
we should _always_ wrap function args in parens. - e.g. `map((mailInfo) => mailInfo.address)`

I'm not super keen on this pattern of a negated, chained map to includes as it's difficult to read.

foladipo replied:
I agree, but the other option I had was to write two or more `if`s, so I decided to stick with this then.

spencern replied:
@foladipo 
I can't fault your decision not to change this pattern as it was the way things were done before. When I'm reviewing code it can be difficult to know what's a new pattern and what's just a few changed lines in a file that has an established pattern (good or bad). I'll make notes on what seems done well or not, but it's not always the fault of the most recent contributor. 

That being said, we need to constantly be attempting to better document our code and this is the job of everyone, not just when we're adding new code.

foladipo replied:
Yeah I agree. A documentation/readability stitch in time probably helps save more than nine.

foladipo commented on imports/plugins/core/accounts/client/templates/profile/profile.html:
```diff
@@ -1,63 +1,69 @@
 <template name="accountProfile">
```
It didn't cross my mind at the time. Besides, this PR was already complex enough, so throwing in a Blaze-template-to-React-component conversion that would need to be tested and code-reviewed was secondary.

foladipo commented on imports/plugins/core/accounts/client/templates/profile/profile.js:
```diff
@@ -88,6 +163,30 @@ Template.accountProfile.helpers({
+      const account = getTargetAccount();
+
+      if (account && account.emails && account.emails[0]) {
+        return account.emails[0].address;
```
I don't understand. Do you mean `account.emails[0].default`?

spencern replied:
@foladipo no, 
account.emails is an array. When we're getting an item out of an array we should be attempting to get the _correct_ item from the array rather than whichever item is in the first position. This is why emails have the `provides` property, and in most cases right now, the `"default"` email is the correct email to use. We should expect that users may have multiple email addresses, and that the email they add first may not be the email they are actively using or that they wish to use for this app.

If you see the code I wrote, it's looking for the email with `provides: "default" rather than just the first element in the array.

foladipo replied:
Oh, I see. I didn't know about this use for the `provides` key in each element. And I do agree that using it as explained above does make more sense.
